lib/watir-screenshot-stitch.rb

```diff
@@ -52,20 +52,31 @@ def base64_geckodriver
     #   @param [Watir::Browser] browser
     # @param [Hash] opts
     #
+    # @return [Thread]
+    #
 
     def save_stitch(path, opts = {})
       return @browser.screenshot.save(path) if base64_capable?
-      @options = opts
-      @path = path
-      calculate_dimensions
+      # wait for anything instance level mutex to become available
+      instance_isolator.synchronize do
+        @options = opts
+        @path = path
+        calculate_dimensions
 
-      return self.save(@path) if (one_shot? || bug_shot?)
+        return self.save(@path) if (one_shot? || bug_shot?)
 
-      build_canvas
-      gather_slices
-      stitch_together
+        build_canvas
+        gather_slices
+        opts[:threaded?] ? Thread.new { stitch }.tap { |t| at_exit { t.join } } : stitch
```

Comment:
"You're right there's a big time savings if you are creating multiple screenshots in the same test." I think the bigger savings will be when stitch at the end of a series of actions on a page then continue from the page. Without a threaded solution the stitch will hold control till everything is finished when watir doesn't need to wait for imagemagick to complete its execution to continue the test especially since watir is about to sit waiting for the next page anyway. The mutex ensures imagemagick will only be stitching a single thread at a time anyway. I think that tact is least likely to break anything and will certainly allow watir to continue on its merry way waiting for stuff to load.

I can imagine someone taking a few stitches than running a synchronous stitch because they need that stitch to return before the next action. Because only one stitch runs at a time running a synchronous stitch also means there are no pending stitch which green lights the next action. I think the rubric here is the principle of least change. If lots of people start using the asynchronous option than I'm sure improvements can be made. I just don't want to make the perfect the enemy of the good here. 

Fair point on the question mark. It looks nice but I agree it's likely bad practice in a public repo. 

I'm not sure what functionality you think is limited by using at exit given the thread is returned and there's no rule against joining it multiple times. If someone wants to join the thread sooner it's not going to break anything. Regarding the alternative of using an afterhook (I assume you mean a watir::afterhook) I'm not sure what action you think should trigger joining all threads. I could see using a 200 page return since that would achieve my primary use case but I'm not sure how that's better than using at_exit

Reply:
If you call `#join`, then move the files somewhere else, the process will error at exit because it can't find the files. I don't think it is unreasonable if a user specifies multithreaded, we can expect them to also control when/how the join happens. I was referring to `Rspec`/`Cucumber` after hooks, where after the test has finished you call join, move the files to where you want to store them, etc.